Stochasically round on conversions from larger types

pjaap opened the pull request:
This resolves #79

This enables the missing stochastic rounding calls on both deterministic and stochastic input.

pjaap committed:
Stochasically round on conversions from larger types

This resolves #79

pjaap commented:
@milankl Thanks for this nice package. I am researching in the area of numerical analysis with stochastic floats and kind of need this small presented feature :)

I think it just needs a change in the few lines. Test are fine locally. Why are the tests not triggered in this PR? I can have a look if you want :)

edit: Workflow needs approval. This just popped up.

milankl commented:
Awesome, thank you!! Will review after CI passes 🥳

milankl commented:
Would you mind adding some tests to verify that this is actually the case?

> - deterministic -> deterministic is deterministically rounded
> - stochastic -> deterministic is deterministically rounded
> - deter -> stoch is stoch rounded
> - stoch -> stoch also stoch rounded

pjaap commented:
Sorry for the late answer. I will add tests and investigate the failure on 1.11. 
I am busy this week, so it may take a few days.

pjaap committed:
Add test for deterministic -> stochastic

pjaap commented:
Testing is not so trivial with non-deterministic values :)

I think
> deterministic -> deterministic is deterministically rounded

is not in the scope of this package.

> stochastic -> deterministic is deterministically rounded

this is already done in the tests. I adjusted the name of the testsets.

> deter -> stoch is stoch rounded

this is now implemented. I check 10000 times that the result differs at least once for 100 random numbers :)

> stoch -> stoch also stoch rounded

I am not sure what to test. The same test as above with `Float64sr`?

giordano commented on test/conversions.jl:
```diff
@@ -45,3 +45,37 @@ end
+
+
+
+for SR in [Float16sr, BFloat16sr, Float32sr]
+
+    # corresponding FP type
+    FP = float(SR)
+
+    @eval begin
+        @testset "Stochastic conversion of Float64 to $(nameof($SR))" begin
```
I don't think `@eval` is necessary.  This could simply be
```suggestion
@testset "Stochastic conversion of Float64 to $(SR)" for SR in (Float16sr, BFloat16sr, Float32sr)
    # corresponding FP type
    FP = float(SR)
    # ...
```
and remove all the dollar signs below.

pjaap replied:
Oh nice. I didn't know that we can add a for loop after the `testset` macro. Will change it in the next push

giordano commented on test/conversions.jl:
```diff
@@ -45,3 +45,37 @@ end
+                    end
+                end
+
+                @test is_stochastic == true
```
No need to check that a boolean is equal to another boolean, you can just test its boolean value 🙂 
```suggestion
                @test is_stochastic
```

pjaap replied:
Yes, I know. I found it more readable as a user in this particular scenario.
A failing test is more semantic since we get the message

```
  Expression: is_stochastic == true
   Evaluated: false == true
```

instead of 
```
  Expression: is_stochastic
```

I think it is better in the first variant.

But I can change it to the second one if you want.

milankl replied:
no need, that's fine by me and I agree with your point

milankl commented:
Just saw https://pretalx.com/juliacon-local-paris-2025/talk/UDQLTD/. Very nice, thanks for talking about this!

Sorry I can have a look at this pull request next week and merge. Feel free to ping me in case I forget

pjaap commented:
Hey there! I was on vacation until today :) so, ping?

I was looking for an interesting topic for the JuliaCon and your package applied to our PDE solvers seemed perfect :)
Luckily, the talk was approved. Looking forward to it.

milankl commented:
This still fails at v1.11 you mind retriggering the CI?

pjaap committed:
Improve formatting a bit

pjaap commented:
I added a small commit. But I cannot start or restart CI.
